src/java.base/share/classes/java/util/ReverseOrderListView.java

```diff
@@ -0,0 +1,381 @@
+     */
+    void checkModifiable() {
+        if (! modifiable)
+            throw new UnsupportedOperationException();
+    }
+
+    class DescendingIterator implements Iterator<E> {
+        final ListIterator<E> it = base.listIterator(base.size());
+        public boolean hasNext() { return it.hasPrevious(); }
+        public E next() { return it.previous(); }
+        public void remove() {
+            checkModifiable();
+            it.remove();
+            // TODO - make sure ListIterator is positioned correctly afterward
+        }
+    }
+
+    class DescendingListIterator implements ListIterator<E> {
+        final ListIterator<E> it;
+
+        DescendingListIterator(int size, int pos) {
+            if (pos < 0 || pos > size)
+                throw new IndexOutOfBoundsException();
+            it = base.listIterator(size - pos);
+        }
+
+        public boolean hasNext() {
+            return it.hasPrevious();
+        }
+
+        public E next() {
+            return it.previous();
+        }
+
+        public boolean hasPrevious() {
+            return it.hasNext();
+        }
+
+        public E previous() {
+            return it.next();
+        }
+
+        public int nextIndex() {
+            return base.size() - it.nextIndex();
+        }
+
+        public int previousIndex() {
+            return nextIndex() - 1;
+        }
+
+        public void remove() {
+            checkModifiable();
+            it.remove();
+        }
+
+        public void set(E e) {
+            checkModifiable();
+            it.set(e);
+        }
+
+        public void add(E e) {
+            checkModifiable();
+            it.add(e);
+            it.previous();
+        }
+    }
+
+    // ========== Iterable ==========
+
+    public void forEach(Consumer<? super E> action) {
+        for (E e : this)
+            action.accept(e);
+    }
+
+    public Iterator<E> iterator() {
+        return new DescendingIterator();
+    }
+
+    public Spliterator<E> spliterator() {
+        // TODO can probably improve this
+        return Spliterators.spliteratorUnknownSize(new DescendingIterator(), 0);
+    }
+
+    // ========== Collection ==========
+
+    public boolean add(E e) {
+        checkModifiable();
+        base.add(0, e);
+        return true;
+    }
+
+    public boolean addAll(Collection<? extends E> c) {
+        checkModifiable();
+        boolean modified = false;
+        for (E e : c) {
+            base.add(0, e);
+            modified = true;
+        }
+        return modified;
+    }
+
+    public void clear() {
+        checkModifiable();
+        base.clear();
+    }
+
+    public boolean contains(Object o) {
+        return base.contains(o);
+    }
+
+    public boolean containsAll(Collection<?> c) {
+        return base.containsAll(c);
+    }
+
+    // copied from AbstractList
+    public boolean equals(Object o) {
+        if (o == this)
+            return true;
+        if (!(o instanceof List))
```

Comment:
should be `if (!(o instanceof List<?> list))`

Reply:
This and similar code are copied from AbstractList. Instead of fiddling around with formatting, I will eventually refactor this into some common support class.